### PR TITLE
Ovt

### PR DESCRIPTION
**Pull request by oliverkurth**
This adds open-vm-tools 9.4.6 and libdnet as a dependency to the boot2docker image, plus a few kernel modules that are recommended for use with vmware.

open-vm-tools and libdnet are built using this little project: https://github.com/vmware/tcl-container. The Dockerfile pulls the binary tarballs from there.

### CODE REVIEW

**Comment by tianon** [comment]
ping @oliverkurth @frapposelli :wink:

**Comment by frapposelli** [comment]
@tianon I believe @oliverkurth could be of more help here, but I'm still watching the PR :wink:

**Comment by tianon** [comment]
@frapposelli ok, fair enough :+1: (thanks for clarifying!)

@oliverkurth I'm really confused here.  Your description mentions "recommended for use with vmware", but @jamtur01 found that the OS won't successfully work without this, and there's no clear information about what this adds that makes it work, or why it doesn't work without this.

**Comment by zeeyang** [comment]
@tianon the cli uses vm tools (ovt) for two things: get guest ip and drop ssh keys (put files). 

As @jamtur01 pointed out, 1 and 2 are cli driver deficiencies. There's a lot of tech debt in the cli driver, and I started a new go lib to address those issues. https://github.com/zeeyang/fusia

I'm not sure what @jamtur01 mean by 3. Why would you need to run virtual box tools inside fusion?

tldr. this ovt patch is needed. cli fusion driver needs more love.

**Comment by jamtur01** [comment]
@zeeyang That should be VMW Tools - they don't auto-start when the VM runs. I had to manually kick them.

**Comment by zeeyang** [comment]
@jamtur01 ah. can you do a quick test with this iso. it will help us pin down if it's your iso build or the vmw environment. https://github.com/cloudnativeapps/boot2docker/releases

**Comment by tianon** [comment]
On 8 January 2015 at 11:02, Zee Yang <notifications@github.com> wrote:
> As @jamtur01 pointed out, 1 and 2 are cli driver deficiencies. There's a lot of tech debt in the cli driver, and I started a new go lib to address those issues. https://github.com/zeeyang/fusia

Couldn't/shouldn't we instead address these deficiencies in Docker
Machine directly so that we don't carry this technical debt forward?

**Comment by SvenDowideit** [comment]
just so its recorded in this PR - #331 shows that the OVT/VMW tools are not needed for b2d to function. if the only reason to add them is to give us a hint how to talk to it, then perhaps the generated but known MAC Address is enough.

But if the tools add something that can't be worked out another way, it would be great to know.

**Comment by oliverkurth** [comment]
Sorry for the late reply - holidays and cold intervened.

So here is a list of things that make this PR valuable:
- the vmballoon driver is used when the host gets low on memory. Basically, the driver will allocate memory in the guest and make it available to the host.
- the vmtoolsd makes it possible to shut down the guest gracefully, by calling 'shutdown' (as opposed to just power off, which is just pulling the power plug). Similarly, it releases dhcp leases when suspending.
You can shutdown/suspend/resume from Fusion/WS or from the cmd line with vmrun (try vmrun -h).
- it makes it possible to query the guest using the 'vmrun' (or 'vmware-vim-cmd' for shared VMs), for example to get  information like the IP adddress and others.
- you can copy files from/to the vm using vmrun

**Comment by SvenDowideit** [comment]
excellent :) thanks for the list. 

I'm only puzzled by the shutdown one - do you mean that vmware doesn't send the ACPI shutdown signal to request a graceful shutdown? mmm, I'm pretty sure it does, but yes, the suspend/resume passthrough is useful - as is anything that helps the vm's time get updated.

**Comment by ehazlett** [comment]
@frapposelli is there anything we can do so get shared folders in here like you have working in the test ISO from cloudnativeapps?  This would also help to get to a single ISO for Docker machine as well. :)

**Comment by tianon** [comment]
Can one of you elaborate on the "technical debt" mentioned above that takes this PR from the realm of "smoothing out VMware issues" into the realm of "required for VMware to even work at all" ?

While all the things this makes better listed above are great, I still don't see any clear-cut case for "this is absolutely required for b2d to function at all in VMware", so I want to understand what actually makes the current ISO not good enough, and what we can do to fix it besides just merging this PR and shoving our "technical debt" that was alluded to above under the rug.  Our rug has plenty of debt underneath it already. :smile:

**Comment by zeeyang** [comment]
@tianon i think @oliverkurth already listed the technical reasons why this patch is needed in a previous comment. The issues @jamtur01 pointed out were cli issues that have been addressed in the docker machine fusion driver.

**Comment by frapposelli** [comment]
@ehazlett the current ISO from cloudnativeapps doesn't have shared folders enabled, it's just a version that contains this patch and the magic string fix.

**Comment by frapposelli** [comment]
As this is becoming fundamental for `docker-machine` I will send out a new PR soon with the build process for `open-vm-tools` added to the boot2docker Dockerfile.

Extensive research proved to be impossible to include Shared Folders functionalities (known as `vmhgfs` internally) in the current TCL 5.x user space (32 bit user space `mount.vmhgfs` and 64 bit kmod don't play well together), hence the roadmap looks like this:

- Boot2Docker based on TCL 5.x will have `open-vm-tools` functionalities but no shared folders.
- Boot2Docker based on TCL 6.x will have full `open-vm-tools` functionalities.

**Comment by tianon** [comment]
> As this is becoming fundamental for `docker-machine` ...

This is the part I'm not getting a good answer to.  What about this pull request or about Docker Machine itself makes this necessary?  The list of things this does/fixes are all good things, but they sound like they're "necessary for Docker Machine support" simply because of some technical debt, so I'm interested what that debt is, and what steps are being taken to pay it back properly.

**Comment by frapposelli** [comment]
@tianon there are several reasons for why `open-vm-tools` are needed for Docker Machine:

- `vmtoolsd` communicates the guest OS IP back to the hypervisor, this functionality is a *must* have for the `vmwarevsphere` as it doesn't work without it. (the driver currently uses a custom boot2docker ISO).

- `vmtoolsd` also enables file exchange between the Guest and the Host, a functionality that will be used to seed the initial SSH keys as we currently rely on B2D user/password for that, [see this comment in the code](https://github.com/docker/machine/blob/master/drivers/vmwarefusion/fusion_darwin.go#L252-L254).

- Once B2D switches to a 64 bit user space we're going to have access to the shared folders functionality  that obviously needs tools running inside the GOS.

There are several other reasons that fall in the technical debt bucket, but these three are the most compelling ones right now.

Sounds reasonable?

**Comment by tianon** [comment]
@frapposelli :heart: :heart:

That's exactly what I was looking for, thanks!  Looking forward to the updated PR. :+1:
